src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.cs

```diff
@@ -202,9 +284,173 @@ public Task LoadAsync()
+        bool IVsProjectRestoreInfoSource.HasPendingNomination
+        {
+            get => CheckIfHasPendingNomination();
+        }
+
+        private bool CheckIfHasPendingNomination()
+        {
+            lock (SyncObject)
+            {
+                Assumes.Present(_project.Services.ActiveConfiguredProjectProvider);
+                Assumes.Present(_project.Services.ActiveConfiguredProjectProvider.ActiveConfiguredProject);
+
+                // Nuget should not wait for projects that failed DTB
+                if (!_enabled || SourceBlock.Completion.IsFaulted || SourceBlock.Completion.IsCompleted)
+                {
+                    return false;
+                }
+
+                // Avoid possible deadlock.
+                // Because RestoreCoreAsync() is called inside a dataflow block it will not be called with new data
+                // until the old task finishes. So, if the project gets nominating restore, it will not get updated data.
+                if (IsNuGetRestoreOnGoing())
+                {
+                    return false;
+                }
+
+                ConfiguredProject? activeConfiguredProject = _project.Services.ActiveConfiguredProjectProvider.ActiveConfiguredProject;
+
+                // After the first nomination, we should check the saved nominated version
+                return IsSavedNominationEmptyOrOlder(activeConfiguredProject);
+            }
+        }
+
+        private bool IsNuGetRestoreOnGoing()
+        {
+            // If NominateForRestoreAsync() has not finished, return false
+            return _nugetRestoreStarted;
+        }
+
+        protected virtual bool IsSavedNominationEmptyOrOlder(ConfiguredProject activeConfiguredProject)
+        {
+            if (!_savedNominatedConfiguredVersion.TryGetValue(activeConfiguredProject.ProjectConfiguration,
+                    out IComparable latestSavedVersionForActiveConfiguredProject) ||
+                activeConfiguredProject.ProjectVersion.IsLaterThan(latestSavedVersionForActiveConfiguredProject))
+            {
+                return true;
+            }
+
+            if (_savedNominatedConfiguredVersion.Count == 1)
+            {
+                return true;
+            }
+
+            foreach (var loadedProject in activeConfiguredProject.UnconfiguredProject.LoadedConfiguredProjects)
+            {
+                if (_savedNominatedConfiguredVersion.TryGetValue(loadedProject.ProjectConfiguration, out IComparable savedProjectVersion))
+                {
+                    if (loadedProject.ProjectVersion.IsLaterThan(savedProjectVersion))
+                    {
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        protected virtual bool IsProjectConfigurationVersionOutOfDate(IReadOnlyCollection<PackageRestoreConfiguredInput>? configuredInputs)
```

Comment:
Why is this nullable?

Reply:
You've used different names for the same thing; IsProjectConfigurationVersionOutOfDate, IsSavedNominationEmptyOrOlder, _savedNominatedConfiguredVersion. Just pick one name for the concept and use it everywhere.

Reply:
perhaps OutOfDate instead of EmptyOrOlder

Reply:
Changed in this [commit](https://github.com/dotnet/project-system/pull/7199/commits/15ec0b30c6d5b6d8b66e16e3fde713a02fe4c6ed)